### PR TITLE
feat: add option to stop validation at first error

### PR DESCRIPTION
**Pull request by ytetsuro**
## Description

Added an option to not detect more than one Error for each item.

## Checklist

- [x] the pull request title describes what this PR does (not a vague title like `Update index.md`)
- [x] the pull request targets the *default* branch of the repository (`develop`)
- [x] the code follows the established code style of the repository
  - `npm run prettier:check` passes
  - `npm run lint:check` passes (An error occurs outside of this PR.)
- [x] tests are added for the changes I made (if any source code was modified)
- [x] documentation added or updated
- [x] I have run the project locally and verified that there are no errors

### Fixes

references #148, references #399, references #188

### CODE REVIEW

**Comment by molinab297-unisys** [comment]
If we could get this merged that would be awesome... #188 shows how much demand there is for this feature.

**Comment by ytetsuro** [comment]
@NoNameProvided

A conflict was occurring, so I rebased it.

However, CI/CD is giving me an error at a point completely unrelated to this fix.
Does this need to be fixed in this PR?

File: src/validation/ValidatorOptions.ts [hunk 1]
```diff
@@ -59,4 +59,9 @@ export interface ValidatorOptions {
    * Settings true will cause fail validation of unknown objects.
    */
   forbidUnknownValues?: boolean;
+
+  /**
+   * Settings true will stop validation when encount first error.
```
Comment (NoNameProvided) [inline src/validation/ValidatorOptions.ts:64]:
```suggestion
   * When set to true, validation will return after encountering the first error. Defaults to false.
```

File: test/functional/validation-options.spec.ts [hunk 2]
```diff
@@ -1086,4 +1086,30 @@ describe('context', () => {
...
+    class MyClass {
+      @IsDefined({
+        message: 'isDefined',
+        context: {
+          bye: 'now',
+        },
+      })
+      @Contains('hello', {
+        message: 'String is not valid. You string must contain a hello word',
+        context: {
```
Comment (NoNameProvided) [inline test/functional/validation-options.spec.ts:1100]: Checking context functionality is not part of this test, please remove and add to where context is tested.

File: test/functional/validation-options.spec.ts [hunk 3]
```diff
@@ -1086,4 +1086,30 @@ describe('context', () => {
...
+    }
+
+    const model = new MyClass();
+    return validator.validate(model, { stopAtFirstError: true }).then(errors => {
+      expect(errors.length).toEqual(1);
+      expect(Object.keys(errors[0].constraints).length).toBe(1);
+      expect(errors[0].contexts['isDefined']).toEqual({ bye: 'now' });
+      expect(errors[0].constraints['isDefined']).toBe('isDefined');
+    });
+  });
```
Comment (NoNameProvided) [inline test/functional/validation-options.spec.ts:1114]: Also, add a test here when `stopAtFirstError` is set to true. (I know this is the default behavior and have a lot of implicit test for this, but let's test it explicitly too.

Comment (ytetsuro) [reply test/functional/validation-options.spec.ts:1114]: I am writing here to test when `stopAtFirst` is set to true.
Are you saying that we should also test for false in this test, as shown below?

```ts
return validator.validate(model, { stopAtFirstError: true }).then(errors => {
      expect(errors.length).toEqual(1);
      expect(Object.keys(errors[0].constraints).length).toBe(1);
      expect(errors[0].contexts['isDefined']).toEqual({ bye: 'now' });
      expect(errors[0].constraints['isDefined']).toBe('isDefined');
    }).then(() => validator.validate(model, { stopAtFirstError: false })).then(errors => {
      expect(Object.keys(errors[0].constraints).length).toBe(2);
    });
```

File: src/validation/ValidatorOptions.ts (hunk 1 above)
Comment (ytetsuro) [reply src/validation/ValidatorOptions.ts]: fixed.

File: test/functional/validation-options.spec.ts (hunk 2 above)
Comment (ytetsuro) [reply test/functional/validation-options.spec.ts]: fixed.

File: test/functional/validation-options.spec.ts (hunk 3 above)
Comment (ytetsuro) [reply test/functional/validation-options.spec.ts:1108]: @NoNameProvided

Comment (NoNameProvided) [reply test/functional/validation-options.spec.ts:1108]: Yes, usually it's good to explicitly test the opposite to show the intent. (We have a lot of implicit test for the default behavior, but it's always better to be explicit. For every feature behind flag there should be a test for on and off version.)

I will go ahead and merge it as it is, in future contribution please keep this in mind. Thanks!

Comment (ytetsuro) [reply test/functional/validation-options.spec.ts:1108]: Yes, sir.

**Comment by ytetsuro** [comment]
Thanks merged.

**Pull request merged**